### PR TITLE
Add format auto-detection for zero-config S3 ingestion

### PR DESCRIPTION
**Pull request by Davidding4718**
### Description

Add a format auto-detection library for Intelligent Ingestion that automatically identifies data format and compression from raw S3 object bytes, eliminating the need for customers to specify a `codec` in their pipeline configuration.

**Key features:**
- Standalone format detection module (`data-prepper-plugins/format-detection/`) with minimal dependencies (snappy-java for decompression)
- Detects compression: gzip, snappy, zstd (via magic bytes)
- Detects binary formats: Parquet, Avro, ORC, PDF, images (via magic bytes)
- Detects text formats: JSON arrays, ND-JSON, CSV, TSV, XML, plain text (via heuristics)
- Handles edge cases: UTF-8 BOM, quoted CSV fields, binary garbage detection, truncated JSON arrays
- Gzip and Snappy decompression for format detection on compressed content

**S3 source integration:**
- When no `codec` is specified in pipeline config, the S3 source enters auto-detect mode
- `AutoDetectS3ObjectWorker` reads first 64KB of each object, detects format, and selects the appropriate codec dynamically
- `AutoDetectCodecFactory` maps detected format to codec plugin (ndjson, json, csv, parquet, avro, newline)
- `DetectionMetrics` emits CloudWatch-compatible metrics via Micrometer with combined format+compression names and per-type duration timers

**Performance:** Average detection time is 1.1ms per file (tested with 5,000 files). Detection overhead is < 0.1% of total pipeline processing time.

---

### Testing

**Unit Tests (57 tests):**
- Compression detection: gzip, zstd, snappy magic bytes
- Binary format detection: Parquet, Avro, ORC, PDF, images
- Text format heuristics: JSON arrays, NDJSON, CSV, TSV, XML, plain text
- Edge cases: empty files, binary garbage, UTF-8 BOM, quoted CSV fields, truncated JSON, large first lines
- Gzip and Snappy decompression integration

**End-to-End Pipeline Testing (S3 → auto-detect → OpenSearch):**

Tested with real S3 bucket using both S3 scan and SQS event-driven modes, with results indexed into an OpenSearch domain.

| File Type | Detected Format | Compression | Confidence | Result |
|-----------|----------------|-------------|------------|--------|
| `.ndjson` | NDJSON | NONE | HIGH | ✅ Fields parsed individually (timestamp, level, service, etc.) |
| `.ndjson.gz` | NDJSON | GZIP | HIGH | ✅ Decompressed + parsed |
| `.ndjson.snappy` | NDJSON | SNAPPY | HIGH | ✅ Decompressed + parsed |
| `.json` (array) | JSON | NONE | HIGH | ✅ Each array element indexed as document |
| `.csv` | CSV | NONE | MEDIUM | ✅ Header row used as field names |
| `.csv.gz` | CSV | GZIP | MEDIUM | ✅ Decompressed + parsed |
| `.csv.snappy` | CSV | SNAPPY | MEDIUM | ✅ Decompressed + parsed |
| `.tsv` | TSV | NONE | MEDIUM | ✅ Tab-delimited parsed |
| `.psv` (pipe) | CSV | NONE | MEDIUM | ✅ Pipe delimiter detected |
| `.parquet` | PARQUET | NONE | HIGH | ✅ Detected and parsed |
| `.pdf` | PDF | NONE | HIGH | ✅ Detected, skipped (unsupported codec) |
| `.log` (plain text) | TEXT | NONE | LOW | ✅ Each line as one event |
| Binary garbage | UNKNOWN | NONE | LOW | ✅ Skipped, not indexed |
| Empty file | UNKNOWN | NONE | LOW | ✅ Skipped |

**Load Test (5,000 files, ~500MB):**
- Detection: 0 failures, average 1.1ms per file, max 3.6ms
- Total detection overhead: 5.6 seconds (< 0.1% of 13 min total pipeline time)
- All 5,000 NDJSON files correctly detected and indexed

**Mixed Format Test (3,600 files — NDJSON + CSV + JSON + compressed variants):**
- All formats correctly detected and routed to appropriate codecs
- Per-type duration metrics confirm compression adds ~1-18ms overhead depending on algorithm
- CloudWatch metrics published successfully under `dataprepper` namespace

---

### CloudWatch Metrics & Observability

Format detection emits metrics via Micrometer (Prometheus + CloudWatch compatible):

**Per format+compression type:**
- `formatDetection.ndjson` / `formatDetection.ndjson.gzip` / `formatDetection.csv.snappy` — count + duration per type
- `formatDetection.total` — aggregate count
- `formatDetection.failed` — unsupported/unknown files
- `formatDetection.confidence.high/medium/low` — confidence breakdown
- `formatDetection.duration` — overall detection timer

**Prometheus endpoint sample** (after 5,000 file load test):
```
s3_auto_detect_pipeline_s3_formatDetection_total 5000.0
s3_auto_detect_pipeline_s3_formatDetection_failed_total 0.0
s3_auto_detect_pipeline_s3_formatDetection_ndjson_total 5000.0
s3_auto_detect_pipeline_s3_formatDetection_duration_seconds_sum 5.631845329
s3_auto_detect_pipeline_s3_formatDetection_confidence_high_total 5000.0
```

**Periodic log summary** (every 60 seconds during activity, with final summary when ingestion stops):

```
╔══════════════════════════════════════════════════╗
║  Format Detection Summary                        ║
╠══════════════════════════════════════════════════╣
║  Total objects processed: 3600                   ║
║  Failed detections:       0                      ║
╠══════════════════════════════════════════════════╣
║  Type Distribution:                              ║
║    ndjson               1000  ( 27.8%)            ║
║    csv                  1000  ( 27.8%)            ║
║    json                 1000  ( 27.8%)            ║
║    ndjson.gzip           200  (  5.6%)            ║
║    csv.gzip              200  (  5.6%)            ║
║    ndjson.snappy         100  (  2.8%)            ║
║    csv.snappy            100  (  2.8%)            ║
╠══════════════════════════════════════════════════╣
║  Confidence:                                     ║
║    HIGH         2300  ( 63.9%)                  ║
║    MEDIUM       1300  ( 36.1%)                  ║
╚══════════════════════════════════════════════════╝
```

---

### Issues Resolved

N/A 

### Check List
- [x] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [x] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/format-detection/src/test/resources/samples/sample.ndjson.gz
Comment (dlvenable) [inline data-prepper-plugins/format-detection/src/test/resources/samples/sample.ndjson.gz:1]: Let's not check in the compressed files. You can use the existing files like `sample.ndjson` and then compress them as part of the test setup.

File: data-prepper-plugins/format-detection/src/test/resources/samples/sample.parquet
Comment (dlvenable) [inline data-prepper-plugins/format-detection/src/test/resources/samples/sample.parquet:1]: Can we generate the Parquet as part of the test setup as well? Binary files are hard to manage in a project.

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/model/configuration/PluginModel.java [hunk 1]
```diff
@@ -213,8 +213,9 @@ public PluginModel deserialize(final JsonParser jsonParser, final Deserializatio
             } else if (jsonParser.currentToken() == JsonToken.VALUE_STRING) {
                 final String value = jsonParser.getValueAsString();
                 if (value.isEmpty()) {
-                    throw context.weirdStringException(value, Map.class,
-                            "Empty string is not allowed for plugin '" + pluginName + "'. Use null, empty (no value), or {} instead.");
+                    // Treat empty string same as null (YAML bare keys like "stdout:" parse as "" in Jackson 2.13+)
+                    isNull = true;
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/org/opensearch/dataprepper/model/configuration/PluginModel.java:217]: It looks like you have some files from other PRs. Please be sure this PR only has the code for auto-detection. You may need to rebase from `upstream/main`.

File: data-prepper-main/build.gradle [hunk 2]
```diff
@@ -52,3 +52,11 @@ artifacts {
     allDependencyJars(jar)
 }
 
+
+task runDataPrepper(type: JavaExec) {
```
Comment (dlvenable) [inline data-prepper-main/build.gradle]: Nice, thanks!

File: data-prepper-plugins/format-detection/src/main/java/org/opensearch/dataprepper/plugins/formatdetection/FormatDetectorPipelineSimulator.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+ *   5. Report what would happen
+ *
+ * Usage:
+ *   java -cp <classpath> FormatDetectorPipelineSimulator /path/to/data/directory
+ */
+public class FormatDetectorPipelineSimulator {
+
+    private static final int SAMPLE_SIZE = 65536;
+
+    public static void main(final String[] args) throws IOException {
```
Comment (dlvenable) [inline data-prepper-plugins/format-detection/src/main/java/org/opensearch/dataprepper/plugins/formatdetection/FormatDetectorPipelineSimulator.java]: Overall, this should provide a new codec rather than a new application.

Also, we should not have another `main()` method in the plugins.

File: data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/S3Source.java [hunk 4]
```diff
@@ -113,16 +117,30 @@ public void start(Buffer<Record<Event>> buffer) {
...
+                        .eventConsumer(eventMetadataModifier)
+                        .s3Client(s3ClientBuilderFactory.getS3Client())
+                        .compressionOption(s3SourceConfig.getCompression())
+                        .build();
+                s3Handler = new S3ObjectWorker(s3ObjectRequest);
+            } else {
+                // Auto-detect mode: no codec specified, detect format per-object
+                LOG.info("No codec configured — using auto-detect mode");
+                final AutoDetectCodecFactory autoDetectCodecFactory = new AutoDetectCodecFactory(pluginFactory);
+                final S3ObjectRequest s3ObjectRequest = s3ObjectRequestBuilder
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/S3Source.java]: Let's not create a split.

A better design would be to include a `Function<S3InputFile, InputCodec>`. Then change the builder to take that instead.

Your code would look more like this:

```
Function<S3InputFile, InputCodec> inputCodecProvider;
if (codecConfiguration != null) {
  inputCodecProvider = new AutoDetectCodecFactory(pluginFactory);
} else {
  final InputCodec codec = pluginFactory.loadPlugin(InputCodec.class, codecPluginSettings);
  inputCodecProvider = s3InputFile -> codec;
}
final S3ObjectRequest s3ObjectRequest = s3ObjectRequestBuilder
  .bucketOwnerProvider(bucketOwnerProvider)
  .codecProvider(inputCodecProvider)
  ...

File: data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/AutoDetectS3ObjectWorker.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * S3ObjectHandler that auto-detects format and compression before processing.
+ * Used when no codec is explicitly configured in the pipeline.
+ */
+class AutoDetectS3ObjectWorker implements S3ObjectHandler {
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/AutoDetectS3ObjectWorker.java]: We should not fork this. The primary difference is that we need to sample the document.

Instead of create a new `S3ObjectHandler` using a `Function<S3InputFile, InputCodec>` should work. I gave some ideas on this below.

**Review by dlvenable — CHANGES_REQUESTED** [review]
Thanks @Davidding4718 for this contribution!

File: data-prepper-plugins/format-detection/src/main/java/org/opensearch/dataprepper/plugins/formatdetection/FormatDetector.java [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+    private static final int CSV_MIN_LINES = 2;
+    private static final int CSV_MAX_SAMPLE_LINES = 10;
+
+    /**
+     * Detect format from a byte array sample.
+     *
+     * @param sample raw bytes from the beginning of the data source
+     * @return detection result with compression, format, and confidence
+     */
+    public FormatDetectionResult detect(final byte[] sample) {
```
Comment (dlvenable) [inline data-prepper-plugins/format-detection/src/main/java/org/opensearch/dataprepper/plugins/formatdetection/FormatDetector.java:58]: We should use Apache Tika to detect formats.

https://tika.apache.org/

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/S3Source.java (hunk 4 above)
Comment (dlvenable) [reply data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/S3Source.java]: I took another look and see that `S3InputFile` has already loaded the object. Maybe we should make this a `Function<S3ObjectReference, InputCodec>`.

File: data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/AutoDetectS3ObjectWorker.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        this.metrics = new DetectionMetrics(pluginMetrics);
+    }
+
+    @Override
+    public void processS3Object(final S3ObjectReference s3ObjectReference,
+                                final S3DataSelection dataSelection,
+                                final AcknowledgementSet acknowledgementSet,
+                                final SourceCoordinator<S3SourceProgressState> sourceCoordinator,
+                                final String partitionKey) throws IOException {
+        final S3InputFile inputFile = new S3InputFile(
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/AutoDetectS3ObjectWorker.java]: We may want to load the object using two requests. Otherwise you've pushed the stream forward. You could use a buffered input stream.

If you do two requests, you should specify a byte range. This may be better anyway to constrain the request length given.
